src/reconcile/rec_child.c

```diff
@@ -35,13 +36,31 @@ __rec_child_deleted(
      * Check visibility. If the truncation is visible to us, we'll also want to know if it's visible
      * to everyone. Use the special-case logic in __wt_page_del_visible to hide prepared truncations
      * as we can't write them to disk.
+     *
+     * We can't write out uncommitted truncations so we need to check the committed flag on the page
+     * delete structure. The committed flag indicates that the truncation has finished being
+     * processed by the transaction commit call and is a separate concept to the visibility which
+     * means that while the truncation may be visible it hasn't finished committing. This can occur
+     * with prepared truncations, which go through two distinct phases in __wt_txn_commit:
+     *   - Firstly the operations on the transaction are walked and the page delete structure has
+     *     its prepare state set to resolved. At this stage the truncate can appear to be visible.
+     *   - After the operations have been resolved the page delete structure is marked as being
+     *     committed.
+     *
+     * Given the order of these operations we must perform the inverse sequence. First check the
+     * committed flag and then check the visibility. There is a concurrency concern here as if the
+     * write to the page delete structure is reordered by the compiler we may see it be set early.
+     * However this is handled by locking the ref in the commit path. Additionally this function
+     * locks the ref. Thus so setting the page delete structure committed flag cannot overlap with
+     * us checking the flag.
```

Comment:
Couple of nitpicks: a comment in the first paragraph ("the visiblity, which") would help with readability.

The CPU can also reorder writes, so I'd probably just say "is reordered".

And "Thus so" reads a little weird, maybe pick one?

Reply:
[5b8a7f7](https://github.com/wiredtiger/wiredtiger/pull/8693/commits/5b8a7f7d9ddfea232f8bb36d6f27210344d8e501)